kernel/sched: Eliminate PRESTART thread state

andyross opened the pull request:
Traditionally threads have been initialized with a PRESTART flag set, which gets cleared when the thread runs for the first time via either its timeout or the k_thread_start() API.

But if you think about it, this is no different, semantically, than SUSPENDED: the thread is prevented from running until the flag is cleared.

So unify the two.  Start threads in the SUSPENDED state, point everyone looking at the PRESTART bit to the SUSPENDED flag, and make k_thread_start() be a synonym for k_thread_resume().

There is some mild code size savings from the eliminated duplication, but the real win here is that we make space in the thread flags byte, which had run out.

andyross committed:
kernel/sched: Eliminate PRESTART thread state

Traditionally threads have been initialized with a PRESTART flag set,
which gets cleared when the thread runs for the first time via either
its timeout or the k_thread_start() API.

But if you think about it, this is no different, semantically, than
SUSPENDED: the thread is prevented from running until the flag is
cleared.

So unify the two.  Start threads in the SUSPENDED state, point
everyone looking at the PRESTART bit to the SUSPENDED flag, and make
k_thread_start() be a synonym for k_thread_resume().

There is some mild code size savings from the eliminated duplication,
but the real win here is that we make space in the thread flags byte,
which had run out.

Signed-off-by: Andy Ross <andyross@google.com>

peter-mitsis commented on kernel/sched.c:
```diff
@@ -632,10 +618,7 @@ void z_sched_wake_thread(struct k_thread *thread, bool is_timeout)
 			if (thread->base.pended_on != NULL) {
 				unpend_thread_no_timeout(thread);
 			}
-			z_mark_thread_as_started(thread);
-			if (is_timeout) {
-				z_mark_thread_as_not_suspended(thread);
-			}
```
I think (for now) we want ...
```
    if (is_timeout) {
        z_mark_thread_as_not_suspend(thread);
    }
```
z_sched_wake_thread() can be called from more than just the timer and we don't necessarily want to resume/unsuspend the thread.

peter-mitsis replied:
Everything else looks great!

andyross replied:
When would we not? I mean, obviously with your SLEEPING patch we'd care about the distinction (and would have to add code here to handle it). But the code as it stands doesn't have that and "wake" and "resume" are synonyms. What's the circumstance where you'd call "wake thread" and expect it not to run?

peter-mitsis replied:
At the present time, z_sched_wake_thread() is also called from the events code with `is_timeout = false` to wake a pended thread. Unconditionally resuming/unsuspending the thread creates an inconsistency when unpending suspended threads that were waiting for a resource. In the case of message queues and semaphores (which do not use z_sched_wake_thread()) the thread may gain the resource but still remain suspended. In the case of events, the suspended thread is resumed/unsuspended upon matching the events.

andyross replied:
> Unconditionally resuming/unsuspending the thread creates an inconsistency when unpending suspended threads that were waiting for a resource.

Right, but you've literally got a patch in the pipeline that is trying to address this inconsistency. :)  Why demand it be fixed ahead of time when we know the disconnect is deeper?

I guess I'm confused: does this change existing behavior or no?  If it breaks something (especially something with a test) it should be fixed.  If it's just us arguing about scheduler philosophy, then my vote is to merge it like this (which seems clearer and safer to me) and then you can come by after this merges and fix it correctly.

andyross replied:
Just to repeat again, because we keep hitting the same disconnect: you and I have **different mental models of what "suspend" is supposed to mean**.  To me, and (most of) the code, "suspend" is a **verb** that means "stop the thread immediately and make sure no further kernel work for it occurs, you can start it again by any means".

You want it to be an **adjective** that means "thread will not run for any reason until resume is called".  And I don't even think that's the wrong design (though it does have size costs, c.f. the extra state bit we're here trying to make room for).  But the code here is AFAICT in keeping with the existing design.  If we're going to change it we should do it in one atomic PR, not piecewise.

andyross committed:
kernel/thread: Fix assumptions in k_thread_state_str()

This table just blindly assumed that the values of _THREAD_xxx state
bits wouldn't change.  That's dangerous.  (That's right, I tried to
change them and broke it.)

Signed-off-by: Andy Ross <andyross@google.com>

andyross commented:
Added an extra patch to clean up the test failures (an earlier version had renumbered the state bits to pack them, but I decided against it to avoid churn as we expect to be adding one back into the hole anyway -- but the gap turned out to be fatal to that table).

andyross commented on subsys/mgmt/mcumgr/transport/src/smp_udp.c:
```diff
@@ -41,9 +41,9 @@ BUILD_ASSERT(0, "Either IPv4 or IPv6 SMP must be enabled for the MCUmgr UDP SMP
 BUILD_ASSERT(sizeof(struct sockaddr) <= CONFIG_MCUMGR_TRANSPORT_NETBUF_USER_DATA_SIZE,
 	     "CONFIG_MCUMGR_TRANSPORT_NETBUF_USER_DATA_SIZE must be >= sizeof(struct sockaddr)");
 
+/* FIXME: dangerous logic, use a kernel API for this */
 #define IS_THREAD_RUNNING(thread)					\
```
It's more that inspecting flags is unlikely to answer the question about state, so I guess it depends what the subsystem wants to know.  The flags tests below look roughly like it wants to know "not dead/aborted/terminated", in which case the proper API would be k_thread_join().
